hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/ObjectEndpoint.java

```diff
@@ -557,8 +559,14 @@ private Response createMultipartKey(String bucket, String key, long length,
             if (range != null) {
               RangeHeader rangeHeader =
                   RangeHeaderParserUtil.parseRangeHeader(range, 0);
-              IOUtils.copyLarge(sourceObject, ozoneOutputStream,
-                  rangeHeader.getStartOffset(),
+              final long skipped =
+                  sourceObject.skip(rangeHeader.getStartOffset());
```

Comment:
skip again uses read and moves the position, why not use like above seek and call copylarge?

Reply:
@bharatviswa504 use skip here due to we are not sure the `sourceObject` is a `seekable` inputstream, so I have to use `skip` method, but luckily, I have override the `skip` method in `KeyInputStream.java`, it look like

```java

  public long skip(long n) throws IOException {
    if (n <= 0) {
      return 0;
    }

    long toSkip = Math.min(n, length - getPos());
    seek(getPos() + toSkip);
    return toSkip;
  }
```

So, for KeyInputStream which is an implementation of InputStream, which implements the skip method to a seek mode, it can bring the better performance.

Reply:
But we use OzoneInputStream here, which has not implemented the skip(). So we use from BaseClass InputStream.


    public long skip(long n) throws IOException {

        long remaining = n;
        int nr;

        if (n <= 0) {
            return 0;
        }

        int size = (int)Math.min(MAX_SKIP_BUFFER_SIZE, remaining);
        byte[] skipBuffer = new byte[size];
        while (remaining > 0) {
            nr = read(skipBuffer, 0, (int)Math.min(size, remaining));
            if (nr < 0) {
                break;
            }
            remaining -= nr;
        }

        return n - remaining;
    }

Reply:
Why can't we add seek to OzoneInputStream like how we have done for S3WrapperInputStream.
As anyway underlying inputStream used in OzoneInputStream is KeyInputStream which has implemented seek.

Reply:
@bharatviswa504 Thanks for your suggestion, indeed, it is necessary to override the skip method of OzoneInputStream, I've done it by the last commit.